<div style="width: 100%; clear: both;">
    <div style="float: left; width: 50%;">
       <img src="http://www.uoc.edu/portal/_resources/common/imatges/marca_UOC/UOC_Masterbrand.jpg", align="left">
    </div>
</div>

<div style="float: right; width: 50%;">
    <p style="margin: 0; text-align:right;">22.503 · Programación para la ciencia de datos</p>
    <p style="margin: 0; text-align:right;">Grado en Ciencia de Datos Aplicada</p>
    <p style="margin: 0; text-align:right; padding-button: 100px;">Estudios de Informática, Multimedia y Telecomunicaciones</p>
</div>

</div>
<div style="width: 100%; clear: both;">
<div style="width:100%;"> </div>

Programación para la ciencia de datos - PEC2
============================

En este Notebook encontraréis un conjunto de ejercicios que conforman la segunda actividad de evaluación continua (PEC) de la asignatura.

Para cada ejercicio, tened en cuenta que:  

* **Es necesario incluir comentarios** de vuestro código, que expliquen cómo ha implementado la solución del problema planteado.  
* **Es imprescindible** citar las referencias consultadas para realizar la actividad. Se valorará que el código proporcionado solucione el problema propuesto y también la calidad del código (comentarios, legibilidad, claridad, uso de las estructuras de datos adecuadas, buena nomenclatura de las variables y funciones, seguimiento del PEP8, etc. ).

Vereis que cada una de las actividades tiene asociada una puntuación, que indica el peso que tiene la actividad sobre la nota de la PEC. Adicionalmente, existe una puntuación asociada a aspectos globales de la PEC (*docstring*, modularidad y estilo), que se evaluarán sobre la totalidad del código entregado.

Además, todas las actividades tienen una etiqueta, que indica los recursos necesarios para llevarla a cabo. Hay tres posibles etiquetas:
* <span style="font-family: Courier New; background-color: #82b74b; color: #000000; padding: 3px; ">SM</span>   **Sólo materiales**: las herramientas necesarias para realizar la actividad se pueden encontrar en los materiales de la asignatura (consideraremos también los materiales de la asignatura Fundamentos de Programación, así como las lecturas obligatorias de material externo que se indican en los notebooks).  
* <span style="font-family: Courier New; background-color: #ffcc5c; color: #000000; padding: 3px; ">EG</span>**Consulta externa guiada**: la actividad puede requerir hacer uso de herramientas que no se encuentran en los materiales de la asignatura, pero el enunciado contiene indicaciones de dónde o cómo encontrar la información adicional necesaria para resolver la actividad.  
* <span style="font-family: Courier New; background-color: #f2ae72; color: #000000; padding: 3px; ">EI</span>**Consulta externa independiente**: la actividad puede requerir hacer uso de herramientas que no se encuentran en los materiales de la asignatura, y el enunciado puede no incluir la descripción de dónde o cómo encontrar esta información adicional. Será necesario que el estudiante busque esta información utilizando los recursos que se han explicado en la asignatura.

Es importante notar que estas etiquetas no indican el nivel de dificultad del ejercicio, sino únicamente la necesidad de consulta de documentación externa para su resolución. Además, recuerde que las **etiquetas son informativas**, pero puede consultar referencias externas siempre que lo desee (aunque no se indique explícitamente) o puede que pueda realizar una actividad sin consultar ningún tipo de documentación. Por ejemplo, para resolver una actividad que sólo requiera los materiales de la asignatura, puede consultar referencias externas si se desea, ya sea tanto para ayudar en la resolución como para ampliar conocimientos!

---

**Importante**: Además de la puntuación de cada ejercicio, que valora la calidad de la solución que propone y los resultados obtenidos, en esta PEC se valoran 3 aspectos adicionales, transversales a todas las actividades:

***Documentación**(**0.25*f puntos**): Todas las funciones de los ejercicios de esta PEC deberán ser correctamente documentadas utilizando docstrings.

***Modularidad**(**0.5*f puntos**): Se valorará la modularidad del código (la correcta estructuración del código en funciones).  

***Estilo**(**0.25*f puntos**): El código debe seguir la guía de estilo de Python [PEP8](https://www.python.org/dev/peps/pep-0008/), exceptuando los casos en los que hacerlo complique la legibilidad del código.

Donde $f$ es la fracción de la PEC realizada. Es decir, la nota de estos aspectos transversales será proporcional a la parte de la PEC realizada.

Así, los aspectos transversales tendrán una puntuación máxima de 1 punto si se entregan todos los ejercicios de la PEC. De lo contrario, la nota de aspectos transversales se multiplicará por la fracción entregada de la PEC . Por ejemplo, si los aspectos transversales están puntuados con un 1 y sólo se ha hecho media PEC, la nota de los aspectos transversales será $1*0.5=0.5$.

La nota de los aspectos transversales se considera incluida en las notas mencionadas para los ejercicios.


---

## Antes de empezar

Por favor, leed atentamente los siguientes párrafos relativos a la originalidad en las actividades de la asignatura antes de empezar a trabajar en la PEC. Si teneis cualquier duda, consultad al profesor colaborador del aula antes de seguir con la actividad.

<div class="alert alert-block alert-success">La falta de originalidad se produce cuando en una actividad aparece contenido que no ha sido elaborado de forma individual por el estudiante y no se referencia adecuadamente su origen. O bien cuando, aunque el contenido externo esté referenciado, éste es tan extenso que no es posible considerar al estudiante el autor o autora de la actividad.
<br /><br />
Así, algunos ejemplos de comportamientos inadecuados son en lo que se refiere a la falta de originalidad son:
<ol>
    <li>Crear la solución de un ejercicio de la PEC en colaboración entre varios estudiantes.</li>
    <li>Incluir un ejercicio de la PEC que utilice código encontrado en internet sin citar la fuente.</li>
    <li>Compartir la solución de la PEC con otros estudiantes de la asignatura.</li>
    </ol>
</div>

## Instrucciones de entrega

Para entregar la actividad es necesario realizar los siguientes pasos:

1. Comprobad que el notebook se ejecuta correctamente en google colab. Es importante que antes de entregar la PEC comprobéis que la versión final del código que se entrega se ejecuta correctamente en su totalidad. Para ello, recomendamos realizar una ejecución completa desde cero del notebook: haga click en *Runtime > Disconnect and Delete Runtime & Run All* y comprobeis que el notebook se ejecuta correctamente.
2. Activad las alertas de estilo y comprobad que el código sigue las convenciones de estilo de Python. Si es necesario, modificad el código para adecuarlo.
3. Confirma que eres el autor único de la PEC y que ésta incluye todas las citas a recursos externos que se hayan utilizado para elaborarla. Añade tu nombre completo al párrafo de la celda siguiente.
4. Para esta PEC sólo será necesario entregar el archivo con el notebook (con extensión .ipynb). Nosotros añadiremos la carpeta "data".
5. Entregad la PEC a través del aula: "Actividades, Entrega de la actividad PEC1".
6. La carga del dataset debe hacerse usando **rutas relativas**, debe cargarse el archivo teniendo en cuenta que el notebook se encuentra en una carpeta que contiene la carpeta data con archivo de datos correspondiente.

<div class="alert alert-block alert-warning">Yo, *Pablo Suárez Reyero*, confirmo que he elaborado de manera individual todas las actividades resueltas en esta PEC y que he incluido las citas en todas las fuentes externas que he utilizado para resolver las actividades. </div>

---

**NOTAS IMPORTANTES:**

1) Durante la PEC contaréis con celdas de test para comprobar que los resultados de vuestro código son correctos. La intención de estas celdas es ayudaros a la resolución de la PEC, pero no se utilizarán directamente en la corrección. En algunos casos poco frecuentes, puede que algún código correcto no pase el test de la celda o que alguno incorrecto si que lo pase. Utilizad el sentido común y consultad a vuestro consultor si tenéis dudas.

2) En varios puntos de la PEC se pide mostrar por pantalla alguna información o estructura de datos. Recordad formatear de forma correcta la salida en todos los casos.

3) Es necesario resolver todos los ejercicios de la PEC evitando recorridos innecesarios de los datos y utilizando, en la medida de lo posible, programación funcional y haciendo un uso avanzado de funciones. **Para obtener la máxima puntuación en los ejercicios es necesario que el código siga estrictamente estos principios además de los aspectos transversales ya mencionados.**

---

In [1]:
# instalemos guia estilo
!pip install pycodestyle pycodestyle_magic
!pip install flake8-nb
%load_ext pycodestyle_magic

In [2]:
# activemos guia estilo
# %pycodestyle_on

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
%cp /content/drive/MyDrive/Colab_Notebooks/activity_2/original.zip ./

# Ejercicio 1

Responded si son Ciertas o Falsas las siguientes preguntas y razonad brevemente vuestra respuesta:

<span style="font-family: Courier New; background-color: #82b74b; color: #000000; padding: 3px; ">NM</span> **(1 punto)**

*(a) Para utilizar una variable global dentro de una función es necesario utilizar la sentencia "nonlocal".*

**Respuesta:** Verdadero, pero habiendo estudiado la teoría, considero que esta pregunta tiene tres respuestas. La primera, es que si lo que se quiere hacer, es solo leer la variable, no hace falta poner ese término. Ahora bien, si lo que se quiere es; escribir/modificar la variable global, con el fin de que el resto de funciones que prosigan lean el nuevo valor, entonces es imprescindible poner ese término. Por último, si se quiere usar una variable local con el mismo nombre que la variable global dentro de una función, entonces no hace falta poner el término, ya que aunque tenga el mismo nombre que la variable global, esta no va a escirbir en la variable global porque no son el mismo tipo de variables.




*(b) Consideramos una variable "float_var" que contiene valores de tipo real.

Una función que tenga por última línea

print(float_var)

devolverá un valor de tipo string. .*

**Respuesta:** Falso, si una función tiene como última linea:
```
def funcion():
  ...
  print(float_var)

```
Entonces significa que la función no está devolviendo ningún valor, ya que no hay ningún ``` return variable```, por lo tanto la función no está devolviendo ningún parámetro, es por esto que lo único que pasará, será que por pantalla, se verá el valor de la variable ```float_var```, y al no haber ningún ```return``` la función devolverá: ```None```.





*(c) Las funciones de Python se pueden pasar como parámetros y ser devueltas por otras funciones*.

**Respuesta:** Verdadero, ya que esto se ha estudiado en teoría. Del mismo modo que una función puede recibir como parámetros, por ejemplo un entero o un string, una función también puede recibir como parámetro otra función. Al igual que a una función se le puede pasar otra función, las funciones también pueden devolver como retorno, una función, al igual que devuelven enteros, tuplas o strings.



(d) Si abrimos un archivo utilizando el modo `w+` mantendremos el contenido del archivo y se generará un error si el archivo no existe.

**Respuesta:** Falso, esto también es algo que se ha estudiado en teoría, y es que, tal y como se ha visto en la **Unidad 3: Archivos e interacción con el sistema**, al abrir un fichero con ```w+``` lo que se va hacer es sobrescribir el contenido del archivo, y en caso de que este no exista, se creará uno nuevo, por lo que el programa no lanzará ningún error al respecto.


## Ejercicio 2

Dadas noticias recientes insinuando que algunos gobiernos ocultan secretos sobre posibles artefactos extraterrestres, el jefe de la empresa de data science donde trabajáis ha decidido crear un proyecto de "fact finding" sobre el tema. Como primer resultado, otros miembros del equipo han conseguido recopilar información sobre avistamientos de OVNIs en Estados Unidos entre 1969 y 2021. Desgraciadamente, estos compañeros no habían cursado esta asignatura y la organización de datos que han hecho no es suficientemente buena.

En este ejercicio utilizaremos los conocimientos sobre gestión de ficheros y programación funcional que estamos aprendiendo para leer la estructura de carpetas existente y transformarla en un dataframe de pandas. En un ejercicio posterior consideraremos otra estructura de archivos que el jefe considera mejor y guardaremos los datos. Durante todo el proceso, estaremos particularmente atentos al espacio ocupado por los archivos.

Hemos recibido los datos en forma de un solo archivo comprimido llamado **original.zip**.

## 2.1

Dado que nos interesa el tamaño que los distintos archivos y carpetas ocupan en memoria, es necesario que **creéis una función para medir el tamaño en archivos o carpetas**. Es necesario que esta función ejecute correctamente tanto si le pasamos la ruta de un archivo como de un directorio. **Descomprimid también el archivo original.zip por código**. **La función debe retornar el tamaño en Megabytes**.

Como prueba, medid el tamaño del archivo **original.zip** y de la estructura de carpetas que se obtiene al descomprimirlo.


## 2.2

Como veréis, la estructura tiene una serie de carpetas con archivos dentro.

En concreto:


```
<original>
    <state_code1>
        <city_name1>
            shape1.txt
            shape2.txt
            ...
            shapeN.txt
        <city_name2>
        ...
    <state_code2>
    ...
        
```

dónde:

- El directorio raíz (llamado "original") contiene una subcarpeta para cada estado de Estados Unidos donde se han detectado avistamientos.
- **state_code** es un código de dos letras para cada uno de los estados de Estados Unidos (por ejemplo CA representa California o AL Alabama).
- Dentro de la carpeta de cada estado existe una carpeta **city_name** con el nombre de cada ciudad donde se han detectado avistamientos. Por ejemplo "Chicago".
- Dentro de cada carpeta de ciudad, habrá un archivo de texto para cada tipo de objeto avistado. Por ejemplo, "circle.txt", "cylinder.txt" o "cigar.txt".
- En cada línea de cada archivo encontrará información (separada por punto y coma ";" referente a los siguientes conceptos):

summary;date_time;duration;text;city_latitude;city_longitude

- Los nombres de estos conceptos no aparecen en los archivos, quedan implícitos por el orden en cada línea. Cada archivo tendrá al menos una línea, pero puede que tenga más.

**Cread una función que reciba como entrada un string con la ruta a una carpeta (conteniendo una estructura como la mencionada) y devuelva un diccionario** con el siguiente formato:

- Las claves del diccionario serán enteros representando el orden en el que se ha encontrado cada una de las líneas de cada uno de los archivos "shapei.txt" mencionados anteriormente. Dependiendo de la implementación puede que leais los archivos en orden diferente.

- El valor asociado a cada clave será, en sí mismo, un diccionario con una clave para cada uno de los siguientes conceptos: summary,city,state,date_time,shape,duration,text,city_latitude,city_longitude y el valor asociado a cada clave según consta en el archivo que se está leyendo.

**Ejemplo:**

En caso de que hubiéramos leído como primer archivo "disk.txt" dentro de la carpeta "IL" y subcarpeta "Chicago", que contiene como primera línea:

"Silver metallic disc observed at Roseland little league baseball field.;1982-06-19 15:00:00;45 seconds;Silver metallic disc observed at Roseland little league baseball field. On a clear...to tell my  story.;41.8841229180932;-87.6674748940178"

Entonces el diccionario daría a esta entrada la clave "0" (primera posición en la secuencia) y tendría, por tanto, como pareja (clave, valor):

(0, {'summary': 'Silver metallic disc observed at Roseland little league baseball field.',
  'city': 'Chicago',
  'state': 'IL',
  'date_time': '1982-06-19 15:00:00',
  'shape': 'disk',
  'duration': '45 seconds',
  'text': "Silver metallic disc observed at Roseland little league baseball field. On a clear...to tell my  story.",
  'city_latitude': '41.8841229180932',
  'city_longitude': '-87.6674748940178'})

Notad que parte de la información (shape, city, state) la hemos obtenido de la estructura de directorios y el resto del archivo "disk.txt". Notad también que hemos omitido parte del campo "text" para mejorar la legibilidad de la PEC.

**importante** Para resolver este apartado (2.2) **No se puede utilizar la librería Pandas.** Su uso resultará en la pérdida de toda la puntuación del apartado.

## 2.3

- Convertid el diccionario obtenido en un dataframe de pandas.
- Guardad este dataframe en un archivo en formato ".csv" llamado **fullDF.csv**.
- Mostrad por pantalla las 10 primeras entradas de este dataframe.
- Medid y mostrad por pantalla el tamaño del archivo **fullDF.csv**.
- Comparad las diferencias en tamaño entre "original.zip", la carpeta descomprimida y "fullDF.csv" utilizando la función del ejercicio 2.1. ¿Cuál ocupa menos espacio. Dado que los tres contienen esencialmente la misma información, ¿cuál es el motivo? Repetid el ejercicio pero ahora hacedlo midiendo el espacio en disco que da el sistema operativo. ¿Existen diferencias? ¿Por qué? Responded a estas preguntas en un pequeño párrafo de texto(<10 lineas).

**Nota:** Para saber los tamaños de archivos "en el sistema operativo" lo más fácil es que copiéis los archivos y carpetas a vuestro drive y los descarguéis para medirlos con las herramientas de vuestro sistema. Esta parte se considerará "extra" y se puede obtener la máxima puntuación en la PEC sin resolverla.

<span style="font-family: Courier New; background-color: #82b74b; color: #000000; padding: 3px; ">NM</span> **(4 puntos)**

**Solución**

Como en esta tarea solo se puede hacer uso del material de la asigntura, la solución que se ha propuesto, hace uso de la recursividad. Esta es la única forma de calcular el tamaño de un directorio que alberga un número indeterminado de carpetas. La recursividad consiste en llamar a la función dentro de ella misma. En este caso, esto nos sirve para poder saber si los elementos que está leyendo, son directorios o ficheros, y consecuentemente, actuar de una manera u otra, dependiendo del caso.


In [5]:
import os
import zipfile as zf
import hashlib

# Respuesta al ejercicio 2.1


def check_file_size(directorio):

    """Esta función devuelve el tamaño del directorio o archivo especificado.

      Esta función acepta un único parámetro de tipo string, correspondiente
      a la ruta que se quiere inspeccionar.

      Argumento de entrada:
        -directorio -- ruta especificada (type = string)

      Variable de retorno:
        -tamano_final: tamaño total del/los elemento·s
    """

    if os.path.isfile(directorio):
        file_size_archivo = os.path.getsize(directorio)
        file_size_archivo = file_size_archivo/(1024**2)
        return file_size_archivo

    else:
        file_size = 0
        elementos = os.listdir(directorio)
        for elemento in elementos:
            directorio_elemento = os.path.join(directorio, elemento)
            file_size += check_file_size(directorio_elemento)
        return file_size

    return 0


# directorio = "/content/drive/MyDrive/Colab_Notebooks/activity_2/original.zip"
directorio = "./original.zip"
size = check_file_size(directorio)

print("El tamaño del fichero en: {} es {} MB\n".format(directorio, size))

# Ahora se comienza con la parte de la decompresión;
print("\n\n Ahora se descomprime el archivo \"original.zip\"")


with zf.ZipFile(directorio, 'r') as zip_f:
    # Mostramos el contenido del zip
    # print(zip_f.printdir())
    # Descomprimimos todo el contenido del zip
    zip_f.extractall("/content")

# Primero cojo la ruta del archivo descomprimido:
directorio_script = os.path.dirname(os.path.abspath("original"))

# Ahora le adjunto el nombre del archivo descomprimido, para luego acceder
# a el correctamente en la función que se ha diseñado antes:
directorio_completo = os.path.join(directorio_script, "original")
print(directorio_completo)

# Se calcula a continuación el tamaño del fichero descomprimido.
size_decompressed = check_file_size(directorio_completo)

print("El tamaño del fichero en: {} es {} MB".format(directorio_completo,
                                                     size_decompressed))

El tamaño del fichero en: ./original.zip es 62.17844772338867 MB



 Ahora se descomprime el archivo "original.zip"
/content/original
El tamaño del fichero en: /content/original es 111.42080402374268 MB


In [6]:
# Test Cell, Public
assert 61.8 <= size <= 62.4, "Wrong Size for original file."
assert 111 <= size_decompressed <= 112, "Wrong Size for decompressed folder."

En la siguiente celda de código, se realiza el segundo apartado, i.e., 2.2


In [7]:
def struct_a_diccionario(ruta):

    """Esta función devuelve el diccionario con el contenido demandado.

      Esta función acepta un único parámetro de tipo string, correspondiente
      a la ruta que se quiere inspeccionar.

      Argumento de entrada:
        -directorio: ruta especificada (type = string)
      Variable de retorno:
        -diccionario: tamaño total del/los elemento·s
    """

    lista_dicts = []
    diccionario_medio = {}
    clau = 0
    with os.scandir(ruta) as estados:
        for estado in estados:
            ruta_estados = os.path.join(estado)
            with os.scandir(ruta_estados) as ciudades:
                for ciudad in ciudades:
                    ruta_ciu = os.path.join(ciudad)
                    with os.scandir(ruta_ciu) as dir_list:
                        for entry in dir_list:
                            if entry.is_file() and entry.name.endswith(".txt"):
                                ruta_txt = os.path.join(ruta_ciu, entry.name)
                                lista_dicts.append(ruta_txt)
                                with open(ruta_txt, 'r') as file:
                                    diccionario = {}
                                    for linea in file:
                                        diccionario[clau] = {}
                                        diccionario[clau]['summary'] = linea.split(';')[0]
                                        diccionario[clau]['city'] = os.path.basename(
                                            os.path.dirname(ruta_txt))
                                        diccionario[clau]['state'] = os.path.basename(
                                            os.path.dirname(ruta_ciu))
                                        diccionario[clau]['date_time'] = linea.split(';')[1]
                                        diccionario[clau]['shape'] = os.path.splitext(entry.name)[0]
                                        diccionario[clau]['duration'] = linea.split(';')[2]
                                        diccionario[clau]['text'] = linea.split(';')[3]
                                        diccionario[clau]['city_latitude'] = linea.split(';')[4]
                                        diccionario[clau]['city_longitude'] = linea.split(';')[5].strip()
                                        diccionario_medio = diccionario_medio | diccionario
                                        clau += 1
    # La función strip lo que hace es eliminar el simbolo del salto de linea
    print('Este es el tamaño del diccionario: ', len(diccionario_medio))
    print('Este es el número de archivos .txt totales: ', len(lista_dicts))
    return diccionario_medio


ruta = './original'
myDict = struct_a_diccionario(ruta)
# help(struct_a_diccionario)

Este es el tamaño del diccionario:  101207
Este es el número de archivos .txt totales:  49543


Este código sirve para visualizar los primeros datos del diccionario (no nos lo pedían, pero era para comprobar que lo he hecho bien)

In [8]:
from itertools import islice


def take(n, iterable):

    """Return the first n items of the iterable as a list."""

    return list(islice(iterable, n))


n_items = take(10, myDict.items())
print(n_items)

[(0, {'summary': 'Cigar shaped object hovering approximately 100 feet in 30 knot wind making no sounds.', 'city': 'Tioga', 'state': 'LA', 'date_time': '1990-05-15 17:50:00', 'shape': 'cigar', 'duration': '10', 'text': 'Cigar shaped object hovering approximately 100 feet in 30 knot wind making no sounds. My fiancé and myself were driving down the road when we saw an object just hovering above the tree tops. We pulled over to a shopping center and watched this object for approximately ten minutes. It had two rows of lights on it, upper and a lower. The lights were going in different directions and would change colures. I estimated the length of it to be around 80 feet. And approximately 25 to 30 feet from top to bottom. I couldn’t tell how deep it was due to seeing it only from the side. This craft was hovering in a 30 to 35 knot wind making no sound. It was barely over the tree tops next to a four lane highway. At the time it was near dusk dark and all you could make out were the shape 

Este código lo he implementado para comprobar el diseño del diccionario que tenía que hacer, es decir, juntando todas las lineas de los archivos txt. Es solo un código de comprobación.

In [9]:
import os


def contar_carpetas(directorio):

    contador = 0
    arxius_txt = []
    total_lineas = 0
    # Recorre recursivamente los subdirectorios
    for raiz, directorios, archivos in os.walk(directorio):
        # Añade las carpetas a la cuenta
        contador += len(directorios)

        elementos = os.listdir(raiz)
        # Filtra solo los archivos con extensión .txt
        for elemento in elementos:
            if elemento.endswith('.txt'):
                arxius_txt.append(elemento)
                ruta_archivo = os.path.join(raiz, elemento)
                with open(ruta_archivo, 'r') as archivo:
                    lineas = archivo.readlines()
                    total_lineas += len(lineas)
    return contador, len(arxius_txt), total_lineas


# Reemplaza 'tu_ruta' con la ruta del directorio que quieres analizar
directorioo = './original'

num_carpetas, num_txt, total_lineas = contar_carpetas(directorioo)
print(f"Número de carpetas en '{directorioo}' y subdirs: {num_carpetas}")
print(f"Número de archivos txt en '{directorioo}': {num_txt}")
print(f"Número de lineas en '{directorioo}': {total_lineas}")

Número de carpetas en './original' y subdirs: 13651
Número de archivos txt en './original': 49543
Número de lineas en './original': 101207


In [10]:
# Test Cell, Public
import random as rn
dict_size = len(myDict)

assert dict_size == 101207, " Incorrect number of entries."
assert len(myDict[0]) == len(myDict[dict_size-1]) == 9, \
 " Incorrect number of entries."

A continuación se transforma el diccionario que se ha creado, en un dataframe. Para ello, en este caso ya tendremos que importar la librería de pandas, para poder utilizar su función

In [11]:
import pandas as pd
import numpy as np
import zipfile as zp
from google.colab import files

# 2.3.1: Ahora se convierte el diccionario obtenido en un df de pandas
df = pd.DataFrame.from_dict(myDict, orient='index')

# 2.3.2 Ahora se guarda el df en un archivo .csv llamado fullDF.csv
df.to_csv('fullDF.csv', index=False)

# 2.3.3 Ahora se muestra por pantalla las 10 primeras entradas del df
print(df.head(10))

# 2.3.4 Ahora se mide y muestra el tamaño (en MB) del archivo fullDF.csv
directorio = './fullDF.csv'
size_df = check_file_size(directorio)
size_prime = os.path.getsize(directorio)/1024**2
print("\nEl tamaño del fichero {} es de {} MB".format(directorio, size_df))
print("El tamaño del fichero {} es de {} MB".format(directorio, size_prime))

# 2.3.5 Ahora se comprueba el tamaño (en MB) de fullDF.csv y de original.zip,
# y se comparan sus diferencias, en el siguiente bloque de texto, depués de las
# siguientes lineas de código, con la función desarrollada en el apartado 2.1

directorio_zip = './original.zip'
directorio_decomp = './original'

size_df = check_file_size(directorio)
size_zip = check_file_size('./original.zip')
size_decomp = check_file_size('./original')

print("\nEl tamaño del fichero {}: {} MB".format(directorio, size_df))
print("El tamaño del fichero {}: {} MB".format(directorio_zip, size_zip))
print("Tamaño del directorio {}: {} MB".format(directorio_decomp, size_decomp))

# files.download('./original')
# files.download('./fullDF.csv')

                                             summary            city state  \
0  Cigar shaped object hovering approximately 100...           Tioga    LA   
1  A dozen bright strars appeared in a straight l...         Crowley    LA   
2                        Machine Sounding Heart Beat         Crowley    LA   
3                             Flying object observed  Denham Springs    LA   
4  An intermittently silver & glowing-white cylin...  Denham Springs    LA   
5                              small flashing lights  Denham Springs    LA   
6  Bright, white, circular light hovering in all ...  Denham Springs    LA   
7                   Bright light hovering in the sky  Denham Springs    LA   
8  Unusual bright lights moving high in the night...  Denham Springs    LA   
9  Glowing green light appeared for a few seconds...  Denham Springs    LA   

             date_time      shape          duration  \
0  1990-05-15 17:50:00      cigar                10   
1  2020-01-07 05:50:00      lig

Respuesta al ejercicio **2.3.5**

En este subapartado del ejercicio **2.3** se nos pedía:

**Comparad las diferencias en tamaño entre "original.zip", la carpeta descomprimida y "fullDF.csv" utilizando la función del ejercicio 2.1. ¿Cuál ocupa menos espacio. Dado que los tres contienen esencialmente la misma información, ¿cuál es el motivo? Repetid el ejercicio pero ahora hacedlo midiendo el espacio en disco que da el sistema operativo. ¿Existen diferencias? ¿Por qué? Responded a estas preguntas en un pequeño párrafo de texto(<10 lineas).**

El archivo **original.zip** pesa menos que el archivo **fullDF.csv** y que la carpeta descomprimida **original**, pues estos pesan **62.178 MB**, **113.576 MB** y **111.421 MB** respectivamente. Era esperable, que **original.zip** fuese más liviano, ya que está comprimido.

Comparando **fullDF.csv** con la carpeta **original**, se ve como es el **fullDF.csv** el que más pesa. Esto puede deberse al método de descompresión aplicado o a la propia estructura del DF, que añade información adicional. No obstante, la diferencia es mínima.

En SO, la carpeta **original** pesa 119,1 MB, el zip **original.zip** pesa 65,2 MB, y el archivo **fullDF.csv** pesa 119,1 MB. Estos datos difieren de los obtenidos con el código, y puede deberse a que el sistema operativo esté añadiendo información adicional relacionada con permisos, o con la gestión de directorios, o incluso también con la compresión con pérdidas y sin pérdidas, aunque el método implementado es el de sin pérdidas.

# Ejercicio 3

En este ejercicio utilizaremos el dataframe construido en el ejercicio anterior para realizar diversas búsquedas en nuestros datos. Además, volveremos a convertir el data frame en otra estructura de archivos y veremos si el resultado mejora el uso del espacio de disco que hemos observado.

**NOTA** Os recomendamos resolver primero completamente el ejercicio 2 antes de intentar el ejercicio 3. En todo caso, disponeis de un pequeño archivo llamado **testDF.csv**. Donde podeis ver una muestra del tipo de registros que debe contener el dataframe resultado del ejercicio 2.

En primer lugar, necesitamos explicitar el año de cada avistamiento:


# 3.1

Añadid una columna "year" al dataframe creado en el ejercicio 2 que contenga el año en que aconteció el avistamiento.


# 3.2

El jefe de la empresa os dice que sabe a ciencia cierta que sólo parte de las entradas del dataframe son evidencia indiscutible de presencia extraterrestas: En concreto deben cumplir, que el campo "text" contenga, a la vez:

- Dos palabras **consecutivas**, la primera que empieze por E y la segunda por T (en mayúsculas)
- Tres palabras **consecutivas**, la primera que empieze por U, la segunda por F y la tercera por O (en este caso, las letras iniciales pueden ser mayúsculas o minúsculas indistintamente)

Por ejemplo, el texto

"...En Tierra las celebraciones son siempre un festival Olímpico, aunque llueve poco..."

Cumpliría ambas condiciones.

Por el contrario, los textos

"...En la Tierra las celebraciones son siempre un festival Olímpico, aunque llueve poco..."

"...En tierra las celebraciones son siempre un festival Olímpico, aunque llueve poco..."

"...En Tierra las celebraciones son siempre un festival de color, aunque llueve poco..."

No lo cumpliría dado que: en el primer caso y el segundo no hay dos palabras consecutivas que empiecen por E y T. En el primer caso las palabras no son consecutivas y en el segundo la E y la T no son ambas mayúsculas. En el tercer caso, si bien la primera condición se cumple, la segunda no.

**Utilizando expresiones regulares**, cread un nuevo dataframe (y llamádlo "df32") que contenga sólo las líneas que cumplan a la vez las dos condiciones mencionadas.


# 3.3

Por motivos de espacio puede ser necesario reducir la información que se guarda para cada registro. Partiendo del dataframe resultado del ejercicio 3.1, cread un nuevo dataframe llamado "df33", en el que habrá eliminado la columna "text".


# 3.4

Cread una funcion que reciba como entrada un string con la ruta a una carpeta y un dataframe y:


**1) cree un diccionario como el que se detalla a continuación**:

- Las claves del diccionario serán enteros representando el orden en el que cada entrada aparece en el dataframe (index).

- El valor asociado a cada clave será, en sí mismo, un diccionario con una clave para cada uno de los siguientes conceptos:

summary,city,state,date_time,shape,duration,text,city_latitude,city_longitude, year

y el valor de cada clave según consta en el dataframe el archivo que se está leyendo.

**Pista: Este parte del ejercicio se puede resolver con una solo instrucción de la librería pandas**

**2) Recorred todas las entradas del diccionario y construid una nueva estructura de ficheros tal y como se detalla a continuación***

- El directorio raíz contendrá una subcarpeta para cada año donde se haya producido un avistamiento, dentro de la carpeta de cada año habrá carpetas para todos los estados donde haya habido avistamientos y dentro de la carpeta de cada estado tendremos una carpeta para cada ciudad en la que haya habido avistamientos. Dentro de cada carpeta de ciudad, habrá un archivo de texto para cada tipo de objeto avistado. Por ejemplo, "circle.txt", "cylinder.txt" o "cigar.txt".

```
<root_folder>
    <year1>
        <state_code1>
            <city_name1>
                shape1.txt
                shape2.txt
                ...
                shapeN.txt
            <city_name2>
            ...
        <state_code2>
        ...
    <year2>
    ...
            
```

- En cada línea de cada fichero deberá escribirse la información referente a los siguientes conceptos (separados por punto y coma ";" ):

summary;date_time;duration;text;city_latitude;city_longitude

- Los nombres de estos conceptos no aparecerán en los archivos y quedarán implícitos en el orden en cada línea. Cada archivo tendrá al menos una línea, pero puede que tenga más.

- En caso de que un concepto no aparezca en un dataframe (por ejemplo "text" en uno de los casos que os pediremos), ignoradlo en la lista de conceptos.

**Ejecutad la funcion creada con los dataframes resultado de los apartados 3.1, 3.2, 3.3.**


# 3.5

- Medid y mostrad por pantalla el tamaño de las tres carpetas resultado del apartado 3.3. Y guardadlas en tres variables llamadas "size_31", "size_32", "size_33".
- Comparad las diferencias entre estas los tamaños de estas 3 carpetas, y el de la carpeta descomprimida "original" según las mide la funcion del apartado 2.1 y según las devuelve vuestro sistema operativo y comentad los resultados en un pequeño párrafo de texto (<10 linies).

**Nota:** Para saber los tamaños de archivos "en el sistema operativo" lo más fácil es que copiéis los archivos y carpetas a vuestro drive y los descarguéis para medirlos con las herramientas de vuestro sistema. Esta parte se considerará "extra" y se puede obtener la máxima puntuación en la PEC sin resolverla.

<span style="font-family: Courier New; background-color: #ffcc5c; color: #000000; padding: 3px; ">EG</span> **(5 puntos)**

**SOLUCIÓN**

 La solución a los apartados **3.1**, **3.2**, **3.3**, **3.4** y **3.5**
 se encuentra en la celda de código de abajo. Mientras que en los siguientes parráfos se encuentran algunas aclaraciones y explicaciones del código.


Para la solución del apartado 3.2 me he inspirado un poco en la siguiente página de StackOverflow, para construir el nuevo dataframe: https://stackoverflow.com/questions/47963548/how-to-append-a-row-to-another-dataframe

Respecto a las expresiones regulares:

- ```'\bE[a-z]*\b \bT[a-z]*\b'``` -> \b especifica el límite de la palabra, [a-z] representan todas las letras en minúscula,
- ```'\b[uU][a-zA-Z]*\b \b[fF][a-zA-Z]*\b \b[oO][a-zA-Z]*\b'``` -> [uU] e igual para f y b, indica que da igual que sea mayúscula o minúscula, que la palabra se va a coger igualmente, por último [a-zA-Z] representan todas las letras en minúscula o mayúscula.


 A la hora de hacer el concat, los indices de las filas copiadas del df antiguo se copiaban tal cual, entonces lo que he hecho, ha sido poner ```ignore_index = True``` para así tener un dataframe con indices lógicos que van de ```0 a len(df)-1```. Esto lo he encontrado en la siguiente página de Stack overflow: https://stackoverflow.com/questions/35528119/pandas-recalculate-index-after-a-concatenation

 Luego, para poder crear el diccionario, a partir del dataframe existente, en el apartado 3.4, se ha hecho uso del siguiente comando: ```df.set_index(df.index).to_dict(orient='index')``` , donde el primer término: ```df.set_index(df.index)``` nos va a permitir crear un diccionario anidado y cuyas claves van a ser los índices de cada una de las filas del data frame. Por último, el término ```.to_dict(orient = 'index')``` va a transformar el DF en un diccionario, y el argumento ```orient = 'index'``` va a permitir tener un diccionario anidado. Referencia: https://stackoverflow.com/questions/26716616/convert-a-pandas-dataframe-to-a-dictionary

Luego, para la creación del sistema de directorios a partir del diccionario creado, he tenido que investigar a cerca de un error a la hora de crear algun directorio, ya que ```os.makedirs(·)``` devolvía un error diciendo que el directorio ya existía, y para solucionarlo, solamente he tenido que poner el flag de ```exist_ok=True```, esto lo he encontrado en la siguiente página web: https://docs.python.org/3/library/os.html#os.makedirs

Por último, en el apartado 3.5, como el campo 'text' no aparece en el dataframe resultante de realizar el apartado 3.3, la función que se ha implementado; ```def dict_and_dir(·)``` no añade el campo ```text```, tal y como se dice en el último punto del enunciado 3.4.

Por último, la comapración de tamaños de las estructuras de datos generados en el punto **3.5** se encuentran debajo de la siguiente celda de código.

In [12]:
import re
import pandas as pd
from itertools import islice
import zipfile as zf

# Esta función la ha creado para poder iterar en el diccionario y comprobarlo


def take(n, iterable):

    """Esta función devuelve los primeros n términos.

      La función acepta dos argumentos, correspondientes a los n primeros
      elementos que se quieren obtener del iterable, y el diccionario que se
      quiere recorrer .

      Argumentos de entrada:
        - n -- "n" primeras posiciones, i.e., type(n) = 'int'
        - iterable -- un diccionario, i.e., type(iterable) = 'dict'

      Variable de retorno:
        - Los "n" primeros elementos del iterable en forma de lista

     """
    return list(islice(iterable, n))

# Esta función es la relativa al apartado 3.4


def dict_and_dir(ruta, df):

    """
    Esta función crea un diccionario que cumple con las instrucciones del
    apartado 3.4, a partir de un df y lo devuelve. Además, esta función también
    se encarga de crear una estructura de ficheros a partir del diccionario
    creado antes, y que cumple con el punto 2 del apartado 3.4.

      La función acepta dos argumentos, correspondientes a los n primeros
      elementos que se quieren obtener del iterable, y el diccionario que se
      quiere recorrer .

      Argumentos de entrada:
        - ruta -- ruta donde se crea el sistema de ficheros, type = 'str'
        - df -- dataframe, type = 'pandas'

      Variable de retorno:
        - Los "n" primeros elementos del iterable en forma de lista
     """

    # Para poder pasarlo todo a un diccionario como antes, se ha de añadir una
    # columna con los indices, pues los indices en el dataframe no tienen
    # columa dedicada asi, se obtendrá el mismo diccionario que aquel creado en
    # el ejercicio 2.2.

    # Primero se crea un diccionario a partir del DF
    diccionario = df.set_index(df.index).to_dict(orient='index')
    n_items = take(10, diccionario.items())
    print(n_items[:10])

    # Ahora se crea el sistema de directorios, recorriendo el diccionario
    # siguiendo las instrucciones del enunciado

    for evento in diccionario.items():
        summ = evento[1]['summary']
        city = evento[1]['city']
        state = evento[1]['state']
        time = evento[1]['date_time']
        shape = evento[1]['shape']
        dur = evento[1]['duration']
        # text = evento[1]['text']
        latitude = evento[1]['city_latitude']
        longitude = evento[1]['city_longitude']
        year = evento[1]['year']

        # Directorio raíz
        year_dir = os.path.join(ruta, year)
        os.makedirs(year_dir, exist_ok=True)

        # Directorio de estados
        state_dir = os.path.join(year_dir, state)
        os.makedirs(state_dir, exist_ok=True)

        # Directorio de ciudades
        city_dir = os.path.join(state_dir, city)
        os.makedirs(city_dir, exist_ok=True)

        # Archivo de texto para los evento
        archivo_objeto = shape+'txt'
        ruta_archivo = os.path.join(city_dir, archivo_objeto)

        # Crear o agregar contenido al archivo
        with open(ruta_archivo, 'w') as file: # En modo 'a' para sobre-escribir
            file.write(f"{summ};{time};{dur};{latitude};{longitude}\n")
    return diccionario


# SOLUCIÓN al apartado 3.1
df['year'] = [year[:4] for year in df['date_time']]

# SOLUCIÓN al apartado 3.2
# Se definen las dos expresiones regulares, para buscar según el criterio
reg1 = r'\bE[a-z]*\b \bT[a-z]*\b'
reg2 = r'\b[uU][a-zA-Z]*\b \b[fF][a-zA-Z]*\b \b[oO][a-zA-Z]*\b'

# Ahora se inicializa el data frame nuevo: df32
df32 = pd.DataFrame()

for subtexto in df['text']:
    if re.findall(reg1, subtexto) and re.findall(reg2, subtexto):
        x = re.findall(reg1, subtexto)
        y = re.findall(reg2, subtexto)
        df32 = pd.concat([df32, df[df['text'] == subtexto]],
                         ignore_index = True)

print(df32.head(len(df32)-1))

# SOLUCIÓN al apartado 3.3
df33 = df.drop('text', axis = 1)

df33.head(len(df33)-1)

# SOLUCIÓN al apartado 3.4
ruta = './root_folder2'
diccionario = dict_and_dir(ruta, df)

# SOLUCIÓN al apartado 3.5
ruta1 = './root31'
ruta2 = './root32'
ruta3 = './root33'

diccionario1 = dict_and_dir(ruta1, df)
diccionario2 = dict_and_dir(ruta2, df32)
diccionario3 = dict_and_dir(ruta3, df33)

# Ahora se calculan los tamaños:
size_31 = check_file_size(ruta1)
size_32 = check_file_size(ruta2)
size_33 = check_file_size(ruta3)

# Ahora se muestra por pantalla los tamaños de los tres directorios
print(f'\nEl directorio {ruta1} pesa: {size_31} MB')
print(f'El directorio {ruta2} pesa: {size_32} MB')
print(f'El directorio {ruta3} pesa: {size_33} MB')

                                             summary            city state  \
0         Security Guard Claims UFO Sighting at Mall      Livingston    NJ   
1  Black Triangle UFO 4 Lights... 1 Red(Blinking)...          Easton    MD   
2  UFO Close Enough To Throw Rock At It,  Makes N...       Bethlehem    PA   
3  7-3-03 A Soundless Vaporless Silver Cylinder O...         Manlius    NY   
4  Very Bright object in morning sky.  ((NUFORC N...   Richmond Hill    GA   
5    Photo of UFO taken in Sedona AZ 3-24-08 4:13 pm          Sedona    AZ   
6  Two triangle shaped green and pink craft seen ...  East Wenatchee    WA   
7  HBCCUFO CANADIAN REPORT:  Update: Light seen a...          Oshawa    ON   
8  Two UFO's Lingering In The Sky On October 31, ...      Washington    DC   

             date_time     shape            duration  \
0  2008-01-17 04:30:00   unknown             Unknown   
1  2011-03-30 01:45:00  triangle       10-15 minutes   
2  1994-06-01 22:00:00  triangle             6-8 mi

El archivo que menos pesa, es```df32 (./root32)``` ya que ese dataframe solo contiene 9 filas de datos, mientras que los otros dos tienen 101207. Comparando estos tamaños con los de la carpeta original, se ve como ```./original``` pesa ```∼111.42``` MB frente a los ```∼11.9``` MB de las nuevas carpetas. Esto puede deberse a que en ```./root31``` y ```./root33``` , ya no esta el campo ```'text'``` y además creo que ahora hay más archivos```.txt``` pero cada uno de ellos solo tiene una linea, mientras que en el directorio ```./original``` habia carpetas con archivos ```.txt``` de 50 lineas. Es por esto, por lo que a lo mejor las nuevas estructuras de datos creadas en este ejercicio pesan menos, respecto a la ```./original```

Ahora comparando el espacio que ocupan las tres carpetas en el SO (varios días despues de haber descargado la carpeta original):

- ```./original``` pesa ```119.1``` MB

Mientras que:

- ```./root31``` pesa ```13.2``` MB
- ```./root32``` pesa ```26``` kB $\equiv$ ```0.026``` MB
- ```./root33``` pesa ```13.2``` MB

Pasa lo mismo, pero puede verse además como el sistema operativo añade bits adicionales, pero la relación de tamaños es la misma, respecto a la comparación de arriba.


In [13]:
# Test Cell, public
assert len(df32) < 30, "Too many entries"

He usado la siguiente celda de código para descargarme las carpetas zip de:

- ```./root31```
- ```./root32```
- ```./root33```

Referencia: https://stackoverflow.com/questions/50453428/how-do-i-download-multiple-files-or-an-entire-folder-from-google-colab

In [14]:
!zip -r './root31.zip' './root31'
!zip -r './root32.zip' './root32'
!zip -r './root33.zip' './root33'

Se han truncado las últimas 5000 líneas del flujo de salida.
updating: root33/2007/MN/Mankato/ (stored 0%)
updating: root33/2007/MN/Mankato/diamondtxt (deflated 23%)
updating: root33/2007/MN/Elk River/ (stored 0%)
updating: root33/2007/MN/Elk River/lighttxt (deflated 14%)
updating: root33/2007/MO/ (stored 0%)
updating: root33/2007/MO/Kansas City/ (stored 0%)
updating: root33/2007/MO/Kansas City/lighttxt (deflated 8%)
updating: root33/2007/MO/Kansas City/triangletxt (deflated 20%)
updating: root33/2007/MO/Kansas City/cylindertxt (deflated 21%)
updating: root33/2007/MO/Kansas City/formationtxt (deflated 18%)
updating: root33/2007/MO/Kansas City/othertxt (deflated 12%)
updating: root33/2007/MO/Cole Camp/ (stored 0%)
updating: root33/2007/MO/Cole Camp/fireballtxt (deflated 18%)
updating: root33/2007/MO/Sweet Springs/ (stored 0%)
updating: root33/2007/MO/Sweet Springs/triangletxt (deflated 17%)
updating: root33/2007/MO/Saint Clair/ (stored 0%)
updating: root33/2007/MO/Saint Clair/flashtxt (